google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,495 @@
+            )
+        if isinstance(status, Exception):
+            status = self._exc_to_status(status)
+        complete_attempt = CompletedAttemptMetric(
+            first_response_latency_ns=self.active_attempt.first_response_latency_ns,
+            duration_ns=time.monotonic_ns() - self.active_attempt.start_time_ns,
+            end_status=status,
+            gfe_latency_ns=self.active_attempt.gfe_latency_ns,
+            application_blocking_time_ns=self.active_attempt.application_blocking_time_ns,
+            backoff_before_attempt_ns=self.active_attempt.backoff_before_attempt_ns,
+            grpc_throttling_time_ns=self.active_attempt.grpc_throttling_time_ns,
+        )
+        self.completed_attempts.append(complete_attempt)
+        self.active_attempt = None
+        for handler in self.handlers:
+            handler.on_attempt_complete(complete_attempt, self)
+
+    def end_with_status(self, status: StatusCode | Exception) -> None:
+        """
+        Called to mark the end of the operation. If there is an active attempt,
+        end_attempt_with_status will be called with the same status.
+
+        Assumes operation is not already in COMPLETED state.
+
+        Causes on_operation_completed to be called for each registered handler.
+
+        Args:
+          - status: The status of the operation.
+        """
+        if self.state == OperationState.COMPLETED:
+            return self._handle_error(
+                INVALID_STATE_ERROR.format("end_with_status", self.state)
+            )
+        final_status = (
+            self._exc_to_status(status) if isinstance(status, Exception) else status
+        )
+        if self.state == OperationState.ACTIVE_ATTEMPT:
+            self.end_attempt_with_status(final_status)
+        self.was_completed = True
+        finalized = CompletedOperationMetric(
+            op_type=self.op_type,
+            completed_attempts=self.completed_attempts,
+            duration_ns=time.monotonic_ns() - self.start_time_ns,
+            final_status=final_status,
+            cluster_id=self.cluster_id or DEFAULT_CLUSTER_ID,
+            zone=self.zone or DEFAULT_ZONE,
+            is_streaming=self.is_streaming,
+            flow_throttling_time_ns=self.flow_throttling_time_ns,
+        )
+        for handler in self.handlers:
+            handler.on_operation_complete(finalized)
+
+    def end_with_success(self):
+        """
+        Called to mark the end of the operation with a successful status.
+
+        Assumes operation is not already in COMPLETED state.
+
+        Causes on_operation_completed to be called for each registered handler.
+        """
+        return self.end_with_status(StatusCode.OK)
+
+    def build_wrapped_predicate(
+        self, inner_predicate: Callable[[Exception], bool]
+    ) -> Callable[[Exception], bool]:
+        """
+        Wrapps a predicate to include metrics tracking. Any call to the resulting predicate
```

Comment:
nit s/Wrapps/Wrap

Im having a hard time wrapping my head around this....whats an example of a predicate that will be wrapped?

Reply:
Looking at the instrumentation pr, it seems like this will be used to wrap the predicate thats passed to the retry infrastructure. This feels a bit hacky...I would expect predicates to be side effect free and this relies on the implementation details of the current retry infrastructure.

I think it would be cleaner to capture retryable attempt results using the on_error_fn param in retries and to capture the overall op status using a try/catch block....it would convey the intent much more clearly